src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+          }
+          Some(samRecord)
+        }
+      }
+      AssignedRecords(sampleOrdinal=sampleOrdinal, records=records)
+    }
+
+    def getBarcodeMetrics: Iterable[SampleBarcodeMetric] = {
+      SampleBarcodeMetric.finalizeMetrics(this.barcodeToMetrics, this.noMatchBarcode)
+      this.barcodeToMetrics.values
+    }
+
+    private def noMatchBarcodeMetric: SampleBarcodeMetric = this.barcodeToMetrics(noMatchBarcode)
+
+    /** Gets the # of mismatches for the given sample and observed sample barcode read bases. */
+    private def sampleOrdinalAndMismatches(sample: Sample, sampleBarcodeReadBases: String): (Int, Int) = {
+      val sampleBarcode = sample.sampleBarcode match {
+        case Some(barcode) => barcode
+        case None => throw new IllegalArgumentException(s"Sample barcode required for sample with id '${sample.sampleId}'")
+      }
+      val forwardMismatches = countMismatches(sampleBarcodeReadBases.getBytes, sampleBarcode.barcodeBytes)
+      val numMismatches = if (matchReverseComplement) {
+        if (!sampleBarcode.isInstanceOf[CustomSampleBarcode]) unreachable("Cannot match the reverse complement unless the barcode is of type 'CustomSampleBarcode'.")
+        val reverseMismatches = countMismatches(sampleBarcodeReadBases.getBytes, sampleBarcode.asInstanceOf[CustomSampleBarcode].reverseComplementedBarcodeBytes)
+        require(reverseMismatches != 0 || forwardMismatches != 0) // otherwise they are reverse complements of each other
+        if (forwardMismatches < reverseMismatches) forwardMismatches
+        else reverseMismatches
+      }
+      else {
+        forwardMismatches
+      }
+      (sample.sampleOrdinal, numMismatches)
+    }
+
+    /** Searches for a matching sample given the observed barcode.  Returns -1 if no match is found, given the
+      * sample sheet and matching parameters */
+    private def getSampleOrdinalFromSampleBarcode(sampleBarcodeReadBases: String, sampleSheet: SampleSheet): Option[Int] = {
+      val numNoCalls: Int = sampleBarcodeReadBases.getBytes.count(base => SequenceUtil.isNoCall(base))
+      val (bestSampleOrdinal, bestMismatches, secondBestMismatches) = if (numNoCalls <= maxNoCalls) {
+        sampleSheet.map { sample => sampleOrdinalAndMismatches(sample, sampleBarcodeReadBases) }
+          .toList.sortBy(_._2).take(2) match {
+            case Nil                              => (-1,           Integer.MAX_VALUE, Integer.MAX_VALUE)
+            case List(bestTuple)                  => (bestTuple._1, bestTuple._2,      Integer.MAX_VALUE)
+            case List(bestTuple, secondBestTuple) => (bestTuple._1, bestTuple._2,      secondBestTuple._2)
+          }
+      }
+      else {
+        (-1, Integer.MAX_VALUE, Integer.MAX_VALUE)
+      }
+      // Make sure we are within the parameter limits and update barcode metrics if necessary
+      if (maxMismatches < bestMismatches || maxNoCalls < numNoCalls || (secondBestMismatches - bestMismatches) < minMismatchDelta) {
```

Comment:
I don't like that this method which sounds like it's just getting/matching, is also updating metrics :/  Maybe it could instead return a `MatchResult` that says a) whether it matched, b) which barcode to and c) with how many mismatches?  And then the updates could happen outside?